### ISSUE TITLE
vm.rollFork not working

### ISSUE THREAD

**Issue by Leonard-Pat** [foundry-rs/foundry#3077]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (1c41585 2022-09-04T09:41:42.826574Z)

### Operating System

Windows

### Describe the bug

I am working in a forked envrionment and as such I am using 
`mainnetFork = vm.createFork(MAINNET_RPC_URL);`
`vm.selectFork(mainnetFork);`
However there seems to be an error when trying to use `vm.rollFork();`
I was using it fine for a couple of weeks however today is started to crash with error 
![image](https://user-images.githubusercontent.com/106159231/188308927-d645324a-e7f8-4c29-923d-2b31d4810621.png)
The way i was using roll fork was like so:
`vm.rollFork(block.number - 15);`
which never gave me an issue. Things ive tried all gave me the same error:
`vm.rollFork(block.number)`
` uint blockNumber = block.number; \\  15471120`
` vm.rollFork(15471120)`
`vm.rollFork(15471120 - 15)`
`vm.rollFork(15471105)`
`uint moveBlocks = block.number - 15`
`vm.rollFork(moveBlocks)`
I have even changed my RPC provider from alchemy to infurar with no avail. Ive updated rust as well as foundry and still nothing. 

Any help would be greatly appreciated! Thank you

**Comment by mattsse** [comment]
we've seen this before, this is an edge case with the depth counter, looks like we're missing a reset somewhere

checking

**Comment by mattsse** [comment]
could you please provide a minimal example test @Leonard-Pat ?
currently failing at reproducing this

**Comment by Leonard-Pat** [comment]
Sure: 
https://gist.github.com/Leonard-Pat/00de1eda51b59b146d7dae072843f080
i removed my RPC provider url but it should still work (well fail...)

**Comment by mattsse** [comment]
hmm I wasn't able to reproduce this here 

https://github.com/foundry-rs/foundry/pull/3079

**Comment by mattsse** [comment]
also tried this manually with `forge test`

```console
forge test                                                                                                                                                                
[⠆] Compiling...
[⠒] Compiling 10 files with 0.8.16
[⠰] Solc 0.8.16 finished in 2.86s
Compiler run successful

Running 1 test for test/Counter.t.sol:testing
[PASS] testFork() (gas: 1245)
Test result: ok. 1 passed; 0 failed; finished in 4.54s
```

which is expected since there are no overflows if built for release

**Comment by Leonard-Pat** [comment]
Yeah very strange, if i open a new project and try it it works, however as soon as i copy all my files across it gives me the same error. Somehow the way ive organised my setup/contracts makes it break ...

**Comment by mattsse** [comment]
I assume you built `forge` manually, because using via `foundryup` would not panic on over/underflows?

but I also wasn't able to reproduce this with the given example with forge built with debug profile

**Comment by Leonard-Pat** [comment]
Yeah i built it using 
`cargo install --git https://github.com/foundry-rs/foundry --profile local --locked foundry-cli anvil`  (if thats what you mean)
Possibly, its just strange how it used to work a week ago

**Comment by mattsse** [comment]
there were some revm changes that landed since then that caused this.

looks like another edge case were we need to adjust the depth counter, but would need a repro to debug and fix this

**Comment by Leonard-Pat** [comment]
Okay well i seemed to fix it if i install WSL and use foundryup :))

**Comment by Leonard-Pat** [comment]
One more problem, sorry i know you cant reproduce these bugs but maybe its still useful to know.
seems like now i cant roll fork fowards after i rolled it back.
`vm.follFork(block.number-15);` this works
`vm.rollFork(block.number + 5);` Now when I move it back forwards it now crashes 
Error:
![image](https://user-images.githubusercontent.com/106159231/188326289-cdcebed7-d745-44cf-8241-03875f898f77.png)
I also built using foundryup this time.

**Comment by mattsse** [comment]
a reproducible example for this would be helpful

**Comment by Leonard-Pat** [comment]
https://github.com/Leonard-Pat/foundrybug
let me know if this works
Main branch - Fails when I have vm.rollFork() inside of a setup function. I have rolled fork back 20 blocks in the previous setup then I move blocks + 1. The token contract is no longer accessible.
fail1 branch - Fails when i have vm.rollFork() outside of setup function and inside my testing contract

**Comment by mattsse** [comment]
sorry facing this:

`[FAIL. Reason: Setup failed: Contract 0xCe71065D4017F316EC606Fe4422e11eB2c47c246 does not exist] setUp() (gas: 0)`


> Main branch - Fails when I have vm.rollFork() inside of a setup function. I have rolled fork back 20 blocks in the previous

what's the error?

we're still taking about the panic mentioned here
https://github.com/foundry-rs/foundry/issues/3077#issuecomment-1236385544 right?

would need some instructions how to properly reproduce this or working example -.-

**Comment by Leonard-Pat** [comment]
Sorry for the confusion basically there are two bugs. You should be able to reproduce the two by simply running `forge test`

First Bug (Main Branch):
Fails setup with reason contract does not exist. 
Bug: 
1. Block was rolled back earlier in the setup file Line 53
2. Token contract deployed in line 55
3.  lLine 85, the fork is rolled forwards 1 Block and attempts to access a function in token contract 
Error: Contract not found. But surely it should be cause it was deployed in an earlier block so should still be visible.

Second Bug (fail1 branch):
Fails with application panic mentioned previously. `Option::unwrap() on a none value`
Bug: 
1. Block was rolled back earlier in the setup file Line 53
2. This occurs when placing vm.rollFork() inside a test function Line 93 
Panic error occurs

Let me know if you can reproduce these

**Comment by mattsse** [comment]
okay,

so the `Contract 0xCe71065D4017F316EC606Fe4422e11eB2c47c246 does not exist` is thrown because the deployed token is not persistent across fork swaps/rolls, this means that the account data is not cloned when we swap/roll a fork.

but you can mark it as such

```solidity
  aaveToken = new AAVE();
  vm.makePersistent(address(aaveToken));
```

the panic was an issue with the depth counter update when we roll fork, which is fixed in #3097

**Comment by Leonard-Pat** [comment]
1) Any idea when fix #3097 will be applied?
2) Thank you so much! the first error is now fixed! ( I'm just curious to why deployed contracts are no longer consistent when a fork is rolled? ,surely that should be the default behaviour since in most cases you'd want that)

p.s I wanted to say that I appreciate all the help/time you have devoted to this

**Comment by mattsse** [comment]
thanks for reporting!

it's now on master and will land in today's nightly in a couple hours
